openedx/core/djangoapps/external_user_ids/docs/decisions/0002-programs-cache.rst

```diff
@@ -0,0 +1,83 @@
+
+* For a large course catalog, this command takes time to run, so there will be at least some time where LMS programs features are broken due to missing data.
+* Site operators may not know that this has to be done. Even if they do, it adds operational complexity.
+* The command may fail.
+
+Recently, the edX.org memcached cluster was restarted due to an increased LMS load. Following the restart, the cache_programs was not run immediately. As a result, there may have been an impact on LMS functionality due to the empty cache.
+
+Developer & operator experience
+=================================
+
+Requiring the regularly-scheduled execution of a management command to make the LMS fully functional adds operational complexity to courses.edx.org, Open edX instances, Devstack, and Sandboxes.
+
+Knowledge & access
+==================
+The requirements to populate the cache via the edx.org Jenkins job, the privileges required or the fact that the cache is not auto refreshed are not well known.
+
+Context boundaries
+==================
+It can be argued that LMS should only have access to data about course-runs, with logic around catalog-courses, curricula, programs, etc. all being handled by the IDAs that are meant to operate within those contexts (Discovery, Registrar, eCommerce, Credentials).
+
+Recommendations
+===============
+1. Developers working on code involving programs (which is not limited Programs theme developers) should be directed to this ADR for context on the usage and limitations of the programs cache.
+2. Consumers of the programs cache should expect cache misses and either handle them gracefully when possible or explicitly fail (with logging).
+    a. Depending on the cache access pattern, it may be impossible to distinguish between a non-existent program and an existent program that is missing from the cache.
+3. Introduce logging to the programs cache, especially in the case of cache misses or an empty cache. This will make debugging easier and make it more apparent when intervention (such as manually running cache_programs) is necessary.
+4. All edX.org developers should be given access to the Jenkins job that runs cache_programs.
+
+Alternatives Considered
+=======================
+* Call into Discovery service if cache miss happens.
+   * Pros:
+      * Removes the issues that arise when a program is not present in the cache.
+      * Handles the case of a completely empty cache.
+   * Cons:
+      * Increases the dependence of LMS on Discovery.
+      * Increases the potential load on the Discovery API.
+      * Involves some engineering effort.
+* Persist Discovery data in LMS MySQL database.
+   * Pros:
+      * Moving to persistent storage would mitigate the “cache accidentally emptied” issue that is inherent to volatile storage.
+
+   * Cons:
+      * Data duplication.
+      * Managing an additional SQL schema.
```

Comment:
The issues that come to my mind here are (a) handling inevitable incoming foreign-key relationships and (b) keeping any uniqueness constraints in sync with the source.

(a) has burned us with CourseOverviews and (b) has burned us with the Credentials programs cache. Both of those are existin SQL based caches.

I'm giving these examples in case you care to elaborate or are questioned about "Managing an additional SQL schema."